implementations/rust/ockam/ockam_node/src/messages.rs

```diff
@@ -124,6 +136,59 @@ pub enum NodeError {
     WorkerExists(Address),
     /// Router already exists
     RouterExists,
+    /// Command rejected
+    Rejected(Reason),
+}
+
+/// The reason why a command was rejected
+#[derive(Debug, Copy, Clone)]
+pub enum Reason {
+    /// Rejected because the node is currently shutting down
+    NodeShutdown,
+    /// Rejected because the worker is currently shutting down
+    WorkerShutdown,
+}
+
+/// Specify the type of node shutdown
+///
+/// For most users `ShutdownType::Graceful()` is recommended.  The
+/// `Default` implementation uses a 1 second timeout.
+#[derive(Debug, Copy, Clone)]
+pub enum ShutdownType {
+    /// Execute a graceful shutdown given a maximum timeout
+    ///
+    /// The following steps will be taken by the internal router
+    /// during graceful shutdown procedure:
+    ///
+    /// * Signal clusterless workers to stop
+    /// * Wait for shutdown ACK hooks from worker set
+    /// * Signal worker clusters in reverse-creation order to stop
+    /// * Wait for shutdown ACK hooks from each cluster before moving onto the
+    ///   next
+    /// * All shutdown-signalled workers may process their entire mailbox,
+    ///   while not allowing new messages to be queued
+    ///
+    /// Graceful shutdown procedure will be pre-maturely terminated
+    /// when reaching the timeout (failover into `Immediate`
+    /// strategy).  **A given timeout of `0` will wait forever!**
```

Comment:
Is that a standard thing in Rust to mean that `0` is `forever` and not `instant`?

Reply:
The reason I didn't choose `core::time::Duration` for this is that in the public API I really wanted to avoid pushing that type on users. But we can have `Option<core::time::Duration>` in our internal representation and then have the API only take values in seconds or milliseconds or something